### DISCUSSION TITLE
How to use multi crazyflie on the lighthouse system.

### DISCUSSION THREAD

**Discussion by fukustar** [bitcraze/discussions#208] [Q&A]
Hi, I am a beginner of using crazyflie.
Now I bought the lighthosue decks and also lighthouse v2. And already finished to use them on cfclient.

So now I want to use lighthouse_openvr_multigrab.py (https://github.com/bitcraze/crazyflie-lib-python/tree/master/examples/lighthouse) for using multi crazyflie. (in my folder, the file name is lighthouse _swarm1-1.py)

But when I tried to use it, there was a problem that

ros2-pc@ROS-PC:~/catkin_ws/src/crazyflie_work$ python3 lighthouse_swarm1-1.py
Opening
Traceback (most recent call last):
 File "lighthouse_swarm1-1.py", line 20, in <module>
  vr = openvr.init(openvr.VRApplication_Other)
 File "/home/ros2-pc/.local/lib/python3.8/site-packages/openvr/__init__.py", line 7209, in init
  initInternal2(applicationType, pStartupInfo)
 File "/home/ros2-pc/.local/lib/python3.8/site-packages/openvr/__init__.py", line 7341, in initInternal2
  openvr.error_code.InitError.check_error_value(error.value) 
 File "/home/ros2-pc/.local/lib/python3.8/site-packages/openvr/error_code/__ini t__.py", line 23, in check_error_value
  raise error_class(error_value, message)
openvr.error_code, InitError_Init_InstallationNotFound

![PXL_20220819_070102148 MP](https://user-images.githubusercontent.com/91782857/185775139-7e632c42-81f8-4629-9636-35a3cb7fc34c.jpg)

I searched that former discussion https://forum.bitcraze.io/viewtopic.php?p=18875&hilit=lighthouse+won%27t+run#p18875
but now I use ubuntu20.04 and also some links that this advicer suggest was 404 not found. So I cannot fix my facing situation.

How to collect this problem? If you have a chip, please give me some advice.

**Comment by knmcguire** [comment]
Hi!

This is pretty old code still originating from the time that the lighthouse system was in early access: https://www.bitcraze.io/2019/03/release-of-the-lighthouse-deck-early-access/. So as I understood correctly, you would like to use a vr controller to control the entire swarm?

If you look at the fault messages, you can see there is something wrong with the openvr installation, which is not installed by standard with the cflib.

**Reply by fukustar** [reply]
Hi, thank you for your reply.

That's correct. I want to control more than two crazyflie with each of position informaiton, so that now I use ighthouse system.

In fact, I already tried to use swarmSequence.py with three crazyflie and collect to fly.
(https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/swarm/swarmSequence.py)
However this file (swarm) has only to fly by oneself in the same time, it cannot fly while interacting with each other (ex. Maintaining a distance of one meter between the two and move forward. ) 

So that I want to use lighthouse because of the limited number of flight methods that can be applied when I don't use it.

**You point out that the code is outdated, but is there any other code for swarm control with lighthouse system that should be used now? I only found the last page. If you know of any, I would like links, etc.**

And also you said that there is something wrong with the openvr installation, so I retry to install, and to check the path of them, but the situation never changed. Now I try to research the reasons of problem.

**Comment by knmcguire** [comment]
Let me try to explain it in a different way

As I understand correctly, you want to use the lighthouse positioning system just for positioning for the entire  swarm, and not use [VR hand controllers ](https://www.vive.com/us/accessory/controller/) to control the swarm. In that sense you should avoid all openVR python scripts as those are old and untested and you don't need openVR (unless you want to use VR hand controllers).

**Lighthouse positioning**
The lighthouse positioning is fully onboard of the crazyflies, so there is no official crazyflie python library script that makes the swarm fly in lighthouse. If you went through [the lighthouse tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-lighthouse/) for one crazyflie and [saved the configuration ](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/lighthouse_tab/#system-management)and uploaded that same config to all the other crazyflies in the swarm (so that they share the same global coordinate system) , then all of them are capable of positioning onboard without any script of the computer. 

**Autonomy**
Since the positioning is calculated all onboard, this means that all autonomy scripts, including the swarm ones, will work fine foryour crazyflies. So if your crazyflies are correctly configured for the lighthouse system using the instructions before, then you can use any of the scripts in the [swarm example folder](https://github.com/bitcraze/crazyflie-lib-python/tree/master/examples/swarm), and your full swarm will be flying with the lighthouse system already. Same for the loco positioning system or the flowdeck, since these scripts don't care about the positioning system used as long as there is one. 

**Links**
These are some good things to read up on:
- Blogpost about differences between different positioning systems: https://www.bitcraze.io/2021/05/positioning-system-overview/
- Detailed information about the lighthouse system: https://www.bitcraze.io/documentation/system/positioning/ligthouse-positioning-system/
- Tutorial for swarm flying: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_swarm_interface/

**Reply by fukustar** [reply]
Thank you four your clear answer.

I don't want to use VR hand controllers, so that I understood to avoid all openVR scripts.
Also, with the lighthouse tutorial, I already tried, but I had missed reading the 'saved the configuration'.

Next work day, I check on the cfclient and tell you about the result.

I really appreciate what you helped.

**Reply by fukustar** [reply]
I’m so sorry to be late reply.

I finally successed to use multi crazyflies on lighthouse system with using saved the configuration. (https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/lighthouse_tab/#system-management)

Thank you so much!

**Reply by krichardsson** [reply]
Great!
